### PR TITLE
Implement cpu_local with GS and ensure GS points to TSS

### PR DESCRIPTION
**Pull request by cqs21**

### CODE REVIEW

**Review by junyang-zh — COMMENTED** [review]
Nice work. I need more time to read them though; seems to maintain `TrapFrame` by our own.

File: ostd/src/arch/x86/trap/gdt.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MPL-2.0 OR MIT
+
+//! Configure Global Descriptor Table (GDT).
+//!
+//! The most contents of this file come from https://github.com/rcore-os/trapframe-rs.
```
Comment (tatetian) [inline ostd/src/arch/x86/trap/gdt.rs:5]: First of all, do not include license or source origin information as Rust doc. Just make them regular comments. Second, it does not say the original code is licensed under MIT. Third, it does not clear which parts are changed by us, which is protected by our copyright license (MPL).

So I suggest changing the file to the following form.

```rust
// The original source code is from [trapframe-rs](https://github.com/rcore-os/trapframe-rs),
// which licensed under the following license.
//
// SPDX-License-Identifier: MIT
//
// Copyright (c) 2020 - 2024 The authors of trapframe-rs 
//
// We make the following new changes:
// * Change 1
// * Change 2
//
// These changes are released under the following license:
//
// SPDX-License-Identifier: MPL-2.0
// 
```

File: ostd/src/arch/x86/trap/gdt.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    USER_SS = sysret + 8;
+    USER_CS = sysret + 16;
+}
+
+// The linker script ensure that cpu_local_tss section is
+// at the beginning of cpu_local, so that gsbase -> LOCAL_TSS.
+#[allow(dead_code)]
+#[link_section = ".cpu_local_tss"]
+static mut LOCAL_TSS: TaskStateSegment = TaskStateSegment::new();
```
Comment (junyang-zh) [inline ostd/src/arch/x86/trap/gdt.rs:76]: The semantics of `.cpu_local_tss` is exactly the same as the current `cpu_local` implementation, you can even define it with:

```rust
cpu_local! {
    static LOCAL_TSS: UnsafeCell<TaskStateSegment> = UnsafeCell:new(TaskStateSegment::new());
}
```

You can use the `LOCAL_TSS.as_ptr()` method to get the address of the task state segment on the current CPU.

Also, use an unsafe cell instead of a static mutable.

File: osdk/src/base_crate/x86_64.ld.template [hunk 3]
```diff
@@ -117,11 +117,14 @@ SECTIONS
     # processor, while it would be copied to other dynamically allocated memory
     # areas for the application processors.
     . = ALIGN(4096);
+    __cpu_local_start = .;
+    .cpu_local_tss          : AT(ADDR(.cpu_local_tss) - KERNEL_VMA) {
```
Comment (junyang-zh) [inline osdk/src/base_crate/x86_64.ld.template:121]: See the comments on the trapframe implementation for the ideas on the task state segment. There's no need for an extra `.cpu_local_tss` segment.

Also if you insist to do so, make the comments better. The above comments just a apply to the `.cpu_local` section. The readers will be confused for the `tss` sectrion.

File: ostd/src/task/mod.rs [hunk 4]
```diff
@@ -53,6 +54,14 @@ impl Task {
         &self.ctx
     }
 
+    /// Sets user fsbase.
+    pub fn set_fsbase(&self, fsbase: usize) {
```
Comment (junyang-zh) [inline ostd/src/task/mod.rs:58]: Rename it to `set_tls_pointer`

File: ostd/src/task/processor.rs [hunk 5]
```diff
@@ -90,6 +92,11 @@ pub(super) fn switch_to_task(next_task: Arc<Task>) {
     //    context and the next task context.
     // 2. The next task context is a valid task context.
     unsafe {
+        let fsbase = (*next_task_ctx_ptr).fsbase;
+        if fsbase != 0 {
+            wrfsbase(fsbase as u64);
```
Comment (junyang-zh) [inline ostd/src/task/processor.rs:97]: Here we should have an abstraction in the architecture dependent module for writing the TLS pointer. This is an architecture-independent module.

File: ostd/src/trap/softirq.rs [hunk 6]
```diff
@@ -96,11 +96,11 @@ impl SoftIrqLine {
 static LINES: Once<[SoftIrqLine; SoftIrqLine::NR_LINES as usize]> = Once::new();
 
 /// Initializes the softirq lines.
```
Comment (junyang-zh) [inline ostd/src/trap/softirq.rs:98]:
```rust
/// Initializes the software IRQ lines if not initialized.
```

File: osdk/src/base_crate/x86_64.ld.template (hunk 3 above)
Comment (cqs21) [reply osdk/src/base_crate/x86_64.ld.template:121]: This segment is intend to make sure that `GSBASE` is pointed to `TSS` struct, so that we don't need to calculate its offset when do `syscall_entry`. This offset could be determined at compile time, but I can't think of a good way to get it.

Comment (junyang-zh) [reply osdk/src/base_crate/x86_64.ld.template:121]: I see. You can leave the implementation as is. But remember to improve the comments.

**Comment by cqs21** [comment]
@boterinas rerun

File: ostd/src/arch/x86/mod.rs [hunk 7]
```diff
@@ -60,7 +60,12 @@ pub(crate) fn check_tdx_init() {
 }
 
 pub(crate) fn init_on_bsp() {
+    // SAFETY: this function is only called once on BSP.
+    unsafe {
+        crate::arch::trap::init(true);
+    }
     irq::init();
+    crate::trap::softirq::init();
```
Comment (tatetian) [inline ostd/src/arch/x86/mod.rs:68]: Is this a bugfix? Why does softirq is enabled on BSPs until this PR of using GS for cpu local storage?

File: ostd/src/trap/softirq.rs [hunk 8]
```diff
@@ -95,12 +95,12 @@ impl SoftIrqLine {
...
 
-/// Initializes the softirq lines.
-///
-/// # Safety
-///
-/// This function must be called only once.
-pub unsafe fn init() {
+/// Initializes the software IRQ lines if not initialized.
+pub fn init() {
+    if LINES.get().is_some() {
```
Comment (tatetian) [inline ostd/src/trap/softirq.rs:100]: What is the motivation of this change?

File: kernel/src/context.rs [hunk 9]
```diff
@@ -29,6 +29,16 @@ impl Context<'_> {
     pub fn get_user_space(&self) -> CurrentUserSpace {
         CurrentUserSpace(self.task.user_space().unwrap().vm_space().clone())
     }
+
+    /// Sets thread-local storage pointer.
+    pub fn set_tls_pointer(&self, tls: usize) {
```
Comment (tatetian) [inline kernel/src/context.rs:34]: I am doubtful about adding `set_tls_pointer` and `tls_pointer` methods for `Context`. They are rarely used. And even without them, the user can still use `ctx.task` to call its `set_tls_pointer` and `tls_pointer` methods directly.

File: ostd/src/arch/x86/cpu/mod.rs [hunk 10]
```diff
@@ -76,6 +77,27 @@ impl UserContext {
...
+        self.set_fsbase(tls)
+    }
+
+    /// Gets thread-local storage pointer.
+    pub fn tls_pointer(&self) -> usize {
+        self.fsbase()
+    }
+
+    /// Loads thread-local storage pointer from fsbase register.
+    pub fn load_tls_pointer(&self) {
```
Comment (tatetian) [inline ostd/src/arch/x86/cpu/mod.rs:92]: This method is confusing. `load_xxx` usually means reading xxx from memory or some other storage medium. But apparently, this means does not load TLS pointer from a source. I suggest calling it `activate_tls_pointer` and revising the Rust doc accordingly.

File: ostd/src/arch/x86/mod.rs (hunk 7 above)
Comment (cqs21) [reply ostd/src/arch/x86/mod.rs:68]: No, I just thought it's more reasonable to init softirq after irq.

File: ostd/src/trap/softirq.rs (hunk 8 above)
Comment (cqs21) [reply ostd/src/trap/softirq.rs:100]: It's not necessarily unsafe.

File: kernel/src/context.rs (hunk 9 above)
Comment (cqs21) [reply kernel/src/context.rs:34]: Right, it's redundant. I'll remove those methods.

File: ostd/src/trap/softirq.rs (hunk 8 above)
Comment (cqs21) [reply ostd/src/trap/softirq.rs]: It has been removed, since it's not related to this PR.

Comment (tatetian) [reply ostd/src/trap/softirq.rs]: I suppose the `unsafe` keyword can be removed. But you do not have to modify the body and guard against calling `init` multiple times (with `if LINES.get().is_some()`). It is a logic error to call `init` more than once. In this case, `Once` will simply panic.

File: ostd/src/arch/x86/mod.rs (hunk 7 above)
Comment (tatetian) [reply ostd/src/arch/x86/mod.rs]: But why is `softirq::init` called in `init_on_bsp`? This means initializing softirq for every CPU.

File: ostd/src/arch/x86/cpu/mod.rs [hunk 11]
```diff
@@ -76,6 +77,27 @@ impl UserContext {
...
+    pub fn set_tls_pointer(&mut self, tls: usize) {
+        self.set_fsbase(tls)
+    }
+
+    /// Gets thread-local storage pointer.
+    pub fn tls_pointer(&self) -> usize {
+        self.fsbase()
+    }
+
+    /// Activates thread-local storage pointer.
```
Comment (junyang-zh) [inline ostd/src/arch/x86/cpu/mod.rs:91]:
```
/// Activates thread-local storage pointer on the current CPU.
///
/// # Safety
///
/// The method by itself is safe because the value of the TLS register won't affect kernel code.
/// But if the user relies on the TLS pointer, make sure that the pointer is correctly set when entering
/// the user space. 
```

File: ostd/src/arch/x86/task/mod.rs [hunk 12]
```diff
@@ -11,15 +13,38 @@ core::arch::global_asm!(include_str!("switch.S"));
...
+
+    /// Activates thread-local storage pointer.
+    pub fn activate_tls_pointer(&self) {
+        if self.fsbase == 0 {
+            return;
+        }
+
+        // SAFETY: fsbase is set by `set_tls_pointer`, we checked above that
+        // it's not zero, so it's valid.
+        unsafe { wrfsbase(self.fsbase as u64) }
```
Comment (junyang-zh) [inline ostd/src/arch/x86/task/mod.rs:46]: Here too. Maybe there's no need for a check.

File: ostd/src/arch/x86/trap/mod.rs [hunk 13]
```diff
@@ -31,14 +48,176 @@ cpu_local_cell! {
...
+    pub fn set_sp(&mut self, sp: usize) {
+        self.general.rsp = sp;
+    }
+
+    /// Get stack pointer.
+    pub fn get_sp(&self) -> usize {
+        self.general.rsp
+    }
+
+    /// Set tls pointer.
```
Comment (junyang-zh) [inline ostd/src/arch/x86/trap/mod.rs:207]: Use plain English in comments. "TLS" all caps or "the thread local storage pointer" for "tls".

File: ostd/src/task/processor.rs [hunk 14]
```diff
@@ -90,6 +90,8 @@ pub(super) fn switch_to_task(next_task: Arc<Task>) {
     //    context and the next task context.
     // 2. The next task context is a valid task context.
     unsafe {
+        (*next_task_ctx_ptr).activate_tls_pointer();
+
         // This function may not return, for example, when the current task exits. So make sure
         // that all variables on the stack can be forgotten without causing resource leakage.
         context_switch(current_task_ctx_ptr, next_task_ctx_ptr);
```
Comment (junyang-zh) [inline ostd/src/task/processor.rs:97]: This is a little bit tricky. Since from the activation of the TLS pointer to the context switch, neither preemption nor IRQs are disabled. So if on SMP preemptive cases the task may be moved to another CPU before the switch happen. Should we `cli` at some place above and call `sti` after loading the registers in the context switch assemby?

Comment (junyang-zh) [reply ostd/src/task/processor.rs:97]: Or maybe activate the TLS pointer in the assembly along with the code loading the registers. Which may be better.

Comment (cqs21) [reply ostd/src/task/processor.rs]: > activate the TLS pointer in the assembly along with the code loading the registers

Done.

**Review by junyang-zh — APPROVED** [review]
Looks decently good to me!

**Comment by tatetian** [comment]
@cqs21 Could you resolve the conflict?

**Comment by cqs21** [comment]
> Could you resolve the conflict?

Resolved

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
